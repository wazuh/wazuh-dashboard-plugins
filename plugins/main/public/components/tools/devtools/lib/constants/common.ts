// Common numeric sentinel values used across DevTools

// Used when string/index searches return no match
export const NOT_FOUND_INDEX = -1;

// Internal timeout status used by network layer (non-HTTP)
export const TIMEOUT_STATUS = -1;
