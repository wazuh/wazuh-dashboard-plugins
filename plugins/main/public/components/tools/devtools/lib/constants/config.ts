/**
 * Non-visual, tweakable settings for DevTools actions.
 */
export const PERMISSIONS_FORBIDDEN_TOKEN = '3029';

// Query parameters that should not be forwarded to the API.
export const RESERVED_QUERY_PARAMS = ['pretty'] as const;
