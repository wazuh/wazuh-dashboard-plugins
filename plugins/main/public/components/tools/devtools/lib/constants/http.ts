/**
 * HTTP-related constants used by DevTools.
 */
export const HTTP_METHODS = ['GET', 'POST', 'PUT', 'DELETE'] as const;
export type HttpMethod = (typeof HTTP_METHODS)[number];

export const DEFAULT_HTTP_METHOD: HttpMethod = 'GET';
