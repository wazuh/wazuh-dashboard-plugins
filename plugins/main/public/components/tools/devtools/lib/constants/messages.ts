/**
 * User-facing messages (centralized for i18n and consistency).
 */
export const MESSAGES = {
  WELCOME: 'Welcome!',
  ERROR_PARSING_JSON: 'Error parsing JSON query',
  INSUFFICIENT_PERMISSIONS: 'Insufficient permissions to execute this method',
  INSUFFICIENT_PERMISSIONS_SHORT: 'Forbidden: insufficient permissions',
} as const;
