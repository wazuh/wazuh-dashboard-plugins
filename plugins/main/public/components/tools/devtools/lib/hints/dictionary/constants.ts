/**
 * UI constants for dictionary hint rendering and labels.
 */
export const HINT_WRAPPER_CLASS = 'wz-hint';
export const HINT_TEXT_CLASS = 'wz-hint__text';
export const HINT_LABEL_CLASS = 'wz-hint__label';

export const LABEL_ENDPOINT = 'endpoint';
export const LABEL_PARAM = 'param';
export const LABEL_FLAG = 'flag';

/**
 * Body-line start matcher for JSON key suggestions.
 * Captures indentation and typed key prefix.
 */
export const BODY_LINE_START_RE = /^(\s*)(?:\"|')(\S*)(?::)?$/;
