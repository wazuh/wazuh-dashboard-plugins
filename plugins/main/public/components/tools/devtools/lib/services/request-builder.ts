import queryString from 'query-string';
import {
  HTTP_METHODS,
  DEFAULT_HTTP_METHOD,
  HttpMethod,
} from '../constants/http';
import type { EditorGroup } from '../types/editor';
import type { BuiltRequest } from '../types/http';
import { safeJsonParse, stripReservedFlags } from '../utils/json';
import { RESERVED_QUERY_PARAMS } from '../constants/config';
import { NOT_FOUND_INDEX } from '../constants/common';

/**
 * Builds sanitized HTTP requests from editor groups.
 * - Detects method and path
 * - Parses inline/body JSON
 * - Strips reserved flags from body and query
 */
export class RequestBuilder {
  build(group: EditorGroup): BuiltRequest {
    const raw = group.requestText || '';
    const method = this.detectMethod(raw);
    const { path, inlineBody } = this.extractPathAndInlineBody(raw, method);

    // Ensure absolute-like path
    const normalizedPath = path
      ? path.startsWith('/')
        ? path
        : `/${path}`
      : '/';

    // Parse and sanitize query parameters
    const { url, query } = queryString.parseUrl(normalizedPath);
    for (const key of RESERVED_QUERY_PARAMS) delete query[key];
    const sanitizedPath = queryString.stringifyUrl(
      { url, query },
      { skipEmptyString: true, skipNull: true },
    );

    // Body: inline has priority, fallback to multi-line JSON body
    let body = safeJsonParse<Record<string, boolean>>(
      (inlineBody as string) || group.requestTextJson,
      {},
    );
    body = stripReservedFlags(body);
    if (typeof body === 'object') body.devTools = true;

    return {
      method,
      path: sanitizedPath,
      body,
    };
  }

  private detectMethod(raw: string): HttpMethod {
    return HTTP_METHODS.find(m => raw.startsWith(m)) || DEFAULT_HTTP_METHOD;
  }

  private extractPathAndInlineBody(
    raw: string,
    method: HttpMethod,
  ): { path: string; inlineBody?: string | false } {
    const withoutMethod = raw.includes(method)
      ? raw.split(method)[1].trim()
      : raw.trim();

    // Inline body handling robust to nested braces: e.g. POST /path { ... { ... } }
    const startIdx = withoutMethod.indexOf('{');
    if (startIdx === NOT_FOUND_INDEX) {
      return { path: withoutMethod };
    }
    // Find matching closing brace using a simple brace counter
    let depth = 0;
    let endIdx = NOT_FOUND_INDEX;
    for (let i = startIdx; i < withoutMethod.length; i++) {
      const ch = withoutMethod[i];
      if (ch === '{') depth++;
      else if (ch === '}') {
        depth--;
        if (depth === 0) {
          endIdx = i;
          break;
        }
      }
    }
    // If we couldn't find a matching closing brace, treat as no inline body
    if (endIdx === NOT_FOUND_INDEX) {
      return { path: withoutMethod };
    }

    const path = withoutMethod.slice(0, startIdx).trim();
    const inlineBody = withoutMethod.slice(startIdx, endIdx + 1);
    return { path, inlineBody };
  }
}
