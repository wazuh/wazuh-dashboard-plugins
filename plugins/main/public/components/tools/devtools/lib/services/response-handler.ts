import { PERMISSIONS_FORBIDDEN_TOKEN } from '../constants/config';

export interface NormalizedResponse {
  body: any;
  status?: number;
  statusText?: string;
  ok: boolean;
}

/**
 * Interprets API responses and extracts consistent metadata.
 */
export class ResponseHandler {
  isPermissionsForbidden(res: any): boolean {
    return typeof res === 'string' && res.includes(PERMISSIONS_FORBIDDEN_TOKEN);
  }

  normalize(res: string | Record<string, any>): NormalizedResponse {
    const response = res || {};
    let status: number | undefined;
    let statusText: string | undefined;
    if (typeof response === 'object') {
      status = response.status;
      statusText = response.statusText;
    }
    const body =
      typeof response === 'object' && 'data' in response
        ? response.data
        : response;
    const hasPayloadError = !!body?.error;
    const ok =
      !hasPayloadError &&
      typeof status === 'number' &&
      status >= 200 &&
      status < 300;

    return { body, status, statusText, ok };
  }
}
