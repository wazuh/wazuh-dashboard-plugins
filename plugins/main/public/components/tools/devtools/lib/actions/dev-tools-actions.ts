import type { EditorLike, EditorOutputLike, SendHooks } from '../types/editor';
import type { HttpClient } from '../types/http';
import { ErrorService } from '../services/error-service';
import { parseErrorForOutput } from '../adapters/error-adapter';
import { GroupingService } from '../grouping';
import { RequestBuilder } from '../services/request-builder';
import { ResponseHandler } from '../services/response-handler';
import { MESSAGES } from '../constants/messages';

/**
 * Encapsulates side-effectful actions for DevTools.
 * Dependencies are injected for testability.
 */
export class DevToolsActions {
  constructor(
    private http: HttpClient,
    private errors = new ErrorService(),
    private requests = new RequestBuilder(),
    private responses = new ResponseHandler(),
    private grouping = new GroupingService(),
  ) {}

  private buildErrorOutput(input: {
    body?: any;
    status?: number;
    statusText?: string;
    fallbackMessage?: string;
  }): { error: string; message: string } {
    const { body, status, statusText, fallbackMessage } = input || {};

    // Build message strictly from HTTP status code and text when available
    let message: string | undefined;
    if (
      typeof status === 'number' &&
      typeof statusText === 'string' &&
      statusText.length
    ) {
      message = `${status} - ${statusText}`;
    } else if (typeof status === 'number') {
      message = String(status);
    } else if (typeof statusText === 'string' && statusText.length) {
      message = statusText;
    } else if (fallbackMessage) {
      message = fallbackMessage;
    }

    // Build error field using API error code when available
    // - If body is an object with `error`, use it as code
    // - Otherwise, fall back to HTTP status code or to fallback message
    let errorPayload: string | undefined;
    if (
      body &&
      typeof body === 'object' &&
      (body as any).error !== undefined &&
      (body as any).error !== null
    ) {
      const maybeCode = (body as any).error;
      errorPayload = String(maybeCode);
    } else if (typeof status === 'number') {
      errorPayload = String(status);
    } else if (fallbackMessage) {
      errorPayload = fallbackMessage;
    }

    return {
      error: errorPayload || 'Unknown',
      message: message || 'Unknown error',
    };
  }

  async send(
    editorInput: EditorLike,
    editorOutput: EditorOutputLike,
    firstTime = false,
    hooks?: SendHooks,
  ) {
    try {
      const groups = this.grouping.parseGroups(editorInput);
      const desiredGroup = this.grouping.selectActiveGroup(
        editorInput,
        firstTime,
        groups,
      );

      if (desiredGroup) {
        const affectedGroups = this.grouping.validateJson(editorInput, groups);
        const hasJsonError = affectedGroups.some(
          item => item === desiredGroup.requestText,
        );
        if (hasJsonError) {
          editorOutput.setValue(MESSAGES.ERROR_PARSING_JSON);
          return;
        }
        const built = this.requests.build(desiredGroup);

        if (!firstTime) {
          const start = Date.now();
          hooks?.onStart?.();
          const response = await this.http.request(
            built.method,
            built.path,
            built.body,
            { returnOriginalResponse: true },
          );

          if (this.responses.isPermissionsForbidden(response)) {
            editorOutput.setValue(MESSAGES.INSUFFICIENT_PERMISSIONS);
            hooks?.onEnd?.({
              status: undefined,
              statusText: MESSAGES.INSUFFICIENT_PERMISSIONS_SHORT,
              durationMs: Date.now() - start,
              ok: false,
            });
            return;
          }

          const { body, status, statusText, ok } =
            this.responses.normalize(response);
          if (ok) {
            editorOutput.setValue(JSON.stringify(body, null, 2));
          } else {
            const out = this.buildErrorOutput({ body, status, statusText });
            editorOutput.setValue(JSON.stringify(out, null, 2));
          }
          hooks?.onEnd?.({
            status,
            statusText,
            durationMs: Date.now() - start,
            ok,
          });
        }
      }

      if (firstTime) editorOutput.setValue(MESSAGES.WELCOME);
    } catch (error: unknown) {
      this.errors.log({ context: 'send', error });
      const resp = (error as any)?.response;
      const status = resp?.status;
      const statusText = resp?.statusText || (error as any)?.message;
      const body = resp?.data;
      const out = this.buildErrorOutput({
        body,
        status,
        statusText,
        fallbackMessage: parseErrorForOutput(error as any),
      });
      editorOutput.setValue(JSON.stringify(out, null, 2));
      hooks?.onEnd?.({ status, statusText, durationMs: 0, ok: false });
    }
  }
}
