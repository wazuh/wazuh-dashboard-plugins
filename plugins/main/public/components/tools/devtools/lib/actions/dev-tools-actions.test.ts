import { DevToolsActions } from './dev-tools-actions';
import { MESSAGES } from '../constants/messages';
import type { EditorOutputLike, SendHooks } from '../types/editor';
import type { HttpClient } from '../types/http';
import { RequestBuilder } from '../services/request-builder';
import { ResponseHandler } from '../services/response-handler';

// Mock only the error adapter to control error string output
jest.mock('../adapters/error-adapter', () => ({
  parseErrorForOutput: jest.fn(() => 'PARSED_ERROR'),
}));

describe('DevToolsActions.send', () => {
  // Minimal editor output stub
  const createOutput = () =>
    ({ setValue: jest.fn() } as unknown as EditorOutputLike);

  // Common default stubs
  const createHttp = (impl?: any) =>
    ({ request: jest.fn(impl) } as unknown as HttpClient & {
      request: jest.Mock;
    });

  const createErrors = () => ({ log: jest.fn() });

  const createRequests = (built: any) => ({
    build: jest.fn(() => built),
  });

  const createResponses = (options?: {
    permissionsForbidden?: boolean;
    normalized?: any;
  }) => ({
    isPermissionsForbidden: jest.fn(() => !!options?.permissionsForbidden),
    normalize: jest.fn(() => options?.normalized ?? {}),
  });

  const createGrouping = (options?: {
    desiredGroup?: any;
    groups?: any[];
    jsonErrors?: string[];
  }) => ({
    parseGroups: jest.fn(() => options?.groups ?? []),
    selectActiveGroup: jest.fn(() => options?.desiredGroup ?? null),
    validateJson: jest.fn(() => options?.jsonErrors ?? []),
  });

  const group = (
    overrides?: Partial<{ requestText: string; requestTextJson: string }>,
  ) => ({
    requestText: 'GET /example',
    requestTextJson: '',
    start: 0,
    end: 0,
    ...(overrides || {}),
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('performs happy path: builds request, calls HTTP, writes JSON, and notifies hooks', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'GET /my/path' });
    const httpResponse = {
      status: 200,
      statusText: 'OK',
      data: { result: 'ok' },
    };

    const http = createHttp(() => Promise.resolve(httpResponse));
    const errors = createErrors();
    const requests = createRequests({
      method: 'GET',
      path: '/my/path',
      body: { a: 1 },
    });
    const responses = createResponses({
      adminForbidden: false,
      normalized: {
        body: httpResponse.data,
        status: 200,
        statusText: 'OK',
        ok: true,
      },
    });
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    // Freeze duration for deterministic assertions
    const nowSpy = jest
      .spyOn(Date, 'now')
      .mockReturnValueOnce(1000) // start
      .mockReturnValueOnce(1100); // end

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    expect(grouping.parseGroups).toHaveBeenCalled();
    expect(grouping.selectActiveGroup).toHaveBeenCalled();
    expect(grouping.validateJson).toHaveBeenCalledWith(expect.anything(), [
      desired,
    ]);
    expect(requests.build).toHaveBeenCalledWith(desired);
    expect(hooks.onStart).toHaveBeenCalledTimes(1);
    expect(http.request).toHaveBeenCalledWith(
      'GET',
      '/my/path',
      { a: 1 },
      { returnOriginalResponse: true },
    );
    expect(responses.isPermissionsForbidden).toHaveBeenCalledWith(httpResponse);
    expect(responses.normalize).toHaveBeenCalledWith(httpResponse);

    // Output should be pretty JSON and stable (snapshot)
    expect((editorOutput.setValue as any).mock.calls[0][0]).toMatchSnapshot(
      'happy-path-output',
    );

    expect(hooks.onEnd).toHaveBeenCalledWith({
      status: 200,
      statusText: 'OK',
      durationMs: 100,
      ok: true,
    });

    nowSpy.mockRestore();
  });

  it('stops early and shows JSON parse error when validation fails', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'POST /broken' });
    const http = createHttp(() => Promise.resolve({}));
    const errors = createErrors();
    const requests = createRequests({
      method: 'POST',
      path: '/broken',
      body: {},
    });
    const responses = createResponses();
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [desired.requestText],
    });

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    expect(editorOutput.setValue).toHaveBeenCalledWith(
      MESSAGES.ERROR_PARSING_JSON,
    );
    expect(http.request).not.toHaveBeenCalled();
    expect(hooks.onStart).not.toHaveBeenCalled();
    expect(hooks.onEnd).not.toHaveBeenCalled();
    expect(requests.build).not.toHaveBeenCalled();
  });

  it('handles insufficient permissions: shows message and calls onEnd with failure', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'DELETE /secure' });
    const httpResponse = 'Forbidden: code 3029'; // PERMISSIONS_FORBIDDEN_TOKEN present
    const http = createHttp(() => Promise.resolve(httpResponse));
    const errors = createErrors();
    const requests = createRequests({
      method: 'DELETE',
      path: '/secure',
      body: {},
    });
    const responses = createResponses({ permissionsForbidden: true });
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    const nowSpy = jest
      .spyOn(Date, 'now')
      .mockReturnValueOnce(2000)
      .mockReturnValueOnce(2150);

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    expect(editorOutput.setValue).toHaveBeenCalledWith(
      MESSAGES.INSUFFICIENT_PERMISSIONS,
    );
    // normalize should not be called when admin forbidden
    expect(responses.normalize).not.toHaveBeenCalled();
    expect(hooks.onEnd).toHaveBeenCalledWith({
      status: undefined,
      statusText: MESSAGES.INSUFFICIENT_PERMISSIONS_SHORT,
      durationMs: 150,
      ok: false,
    });

    nowSpy.mockRestore();
  });

  it('firstTime=true shows welcome and skips network even when a group is selected', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'GET /ignored' });
    const http = createHttp(() => Promise.resolve({}));
    const errors = createErrors();
    const requests = createRequests({
      method: 'GET',
      path: '/ignored',
      body: {},
    });
    const responses = createResponses();
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, true, hooks);

    expect(editorOutput.setValue).toHaveBeenCalledWith(MESSAGES.WELCOME);
    expect(http.request).not.toHaveBeenCalled();
    expect(hooks.onStart).not.toHaveBeenCalled();
    expect(hooks.onEnd).not.toHaveBeenCalled();
  });

  it('handles thrown HTTP errors: logs, renders error JSON {error, message}, and calls onEnd with error metadata', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'GET /boom' });
    const boom = Object.assign(new Error('Boom'), {
      response: { status: 500, statusText: 'Internal' },
    });

    const http = createHttp(() => Promise.reject(boom));
    const errors = createErrors();
    const requests = createRequests({ method: 'GET', path: '/boom', body: {} });
    const responses = createResponses();
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    expect(errors.log).toHaveBeenCalledWith({ context: 'send', error: boom });
    expect(editorOutput.setValue).toHaveBeenCalledWith(
      JSON.stringify({ error: '500', message: '500 - Internal' }, null, 2),
    );
    expect(hooks.onEnd).toHaveBeenCalledWith({
      status: 500,
      statusText: 'Internal',
      durationMs: 0,
      ok: false,
    });
  });

  it('renders error output using payload error code and HTTP status text when response body has error', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    const desired = group({ requestText: 'GET /notfound' });
    const httpResponse = {
      status: 404,
      statusText: 'Not Found',
      data: { error: 3013, message: 'Some API message' },
    };

    const http = createHttp(() => Promise.resolve(httpResponse));
    const errors = createErrors();
    const requests = createRequests({
      method: 'GET',
      path: '/notfound',
      body: {},
    });
    const responses = createResponses({
      normalized: {
        body: httpResponse.data,
        status: 404,
        statusText: 'Not Found',
        ok: false,
      },
    });
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests as any,
      responses as any,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    expect(editorOutput.setValue).toHaveBeenCalledWith(
      JSON.stringify({ error: '3013', message: '404 - Not Found' }, null, 2),
    );
    expect(hooks.onEnd).toHaveBeenCalledWith({
      status: 404,
      statusText: 'Not Found',
      durationMs: expect.any(Number),
      ok: false,
    });
  });

  it('integration: uses real RequestBuilder and ResponseHandler to shape request/response', async () => {
    const editorOutput = createOutput();
    const hooks: SendHooks = { onStart: jest.fn(), onEnd: jest.fn() };

    // Inline body + reserved flags in body/query => sanitized by RequestBuilder
    const desired = group({
      requestText: 'POST agents?pretty=true {"pretty": true, "x": 1}',
    });

    let capturedArgs: any[] | null = null;
    const http = createHttp(
      (method: any, path: any, body: any, options: any) => {
        capturedArgs = [method, path, body, options];
        return Promise.resolve({
          status: 201,
          statusText: 'Created',
          data: { id: 'abc' },
        });
      },
    );

    const errors = createErrors();
    const requests = new RequestBuilder();
    const responses = new ResponseHandler();
    const grouping = createGrouping({
      desiredGroup: desired,
      groups: [desired],
      jsonErrors: [],
    });

    const nowSpy = jest
      .spyOn(Date, 'now')
      .mockReturnValueOnce(3000)
      .mockReturnValueOnce(3200);

    const actions = new DevToolsActions(
      http,
      errors as any,
      requests,
      responses,
      grouping as any,
    );

    await actions.send({} as any, editorOutput, false, hooks);

    // RequestBuilder should purge reserved query and body flags, and mark devTools=true
    expect(capturedArgs).toEqual([
      'POST',
      '/agents',
      { x: 1, devTools: true },
      { returnOriginalResponse: true },
    ]);

    // ResponseHandler should pretty-print the response body
    expect(editorOutput.setValue).toHaveBeenCalledWith(
      JSON.stringify({ id: 'abc' }, null, 2),
    );
    expect(hooks.onEnd).toHaveBeenCalledWith({
      status: 201,
      statusText: 'Created',
      durationMs: 200,
      ok: true,
    });

    nowSpy.mockRestore();
  });
});
