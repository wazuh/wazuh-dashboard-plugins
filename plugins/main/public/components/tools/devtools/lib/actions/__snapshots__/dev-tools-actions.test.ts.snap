// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevToolsActions.send performs happy path: builds request, calls HTTP, writes JSON, and notifies hooks: happy-path-output 1`] = `
"{
  \\"result\\": \\"ok\\"
}"
`;
